File: static/merchant_pds_format_v1.js
/**
 * Product Design System Compliance V1 — shared merchant presentation helpers.
 * Presentation only: currency format + governed language sanitization.
 */
(function () {
  "use strict";

  var SAR_PREFIX = "SR\u00a0";

  function formatMerchantSar(amount) {
    var n = Math.round(parseFloat(String(amount == null ? "" : amount).replace(/,/g, "")) || 0);
    if (isNaN(n)) return "";
    return SAR_PREFIX + n.toLocaleString("en-US");
  }

  function formatMerchantSarHtml(amount) {
    var text = formatMerchantSar(amount);
    if (!text) return "";
    return (
      '<span class="cf-currency-atom cftyp-currency" data-cf-currency="1" dir="ltr">' +
      text +
      "</span>"
    );
  }

  /** Client-side guard for rejected Merchant Language strings (display layer). */
  function sanitizeMerchantLanguage(text) {
    var s = String(text == null ? "" : text).trim();
    if (!s) return "";
    if (/اكتملت\s*مشتريات/.test(s)) return "عادوا وأكملوا الشراء";
    return s;
  }

  window.formatMerchantSar = formatMerchantSar;
  window.formatMerchantSarHtml = formatMerchantSarHtml;
  window.sanitizeMerchantLanguage = sanitizeMerchantLanguage;
})();

File: services/merchant_pds_format_v1.py
# -*- coding: utf-8 -*-
"""Merchant PDS format helpers — certified compact currency (presentation only)."""
from __future__ import annotations

from markupsafe import Markup, escape


def _merchant_sar_number(amount) -> int | None:
    if amount is None:
        return None
    if isinstance(amount, str):
        cleaned = amount.replace(",", "").replace("٬", "").strip()
        if not cleaned or cleaned in {"—", "-"}:
            return None
        if cleaned in {"…", "..."}:
            return None
        try:
            return round(float(cleaned))
        except ValueError:
            return None
    try:
        return round(float(amount))
    except (TypeError, ValueError):
        return None


def format_merchant_sar(amount) -> str:
    """Certified compact dashboard currency: ``SR 449``."""
    if isinstance(amount, str) and amount.strip() in {"…", "..."}:
        return amount.strip()
    n = _merchant_sar_number(amount)
    if n is None:
        return ""
    return f"SR {n:,}"


def format_merchant_sar_html(amount) -> Markup:
    text = format_merchant_sar(amount)
    if not text:
        return Markup("")
    return Markup(
        '<span class="cf-currency-atom cftyp-currency" data-cf-currency="1" dir="ltr">'
        f"{escape(text)}"
        "</span>"
    )
